python/mxnet/pretrained_model.py

```diff
@@ -0,0 +1,576 @@
+        >>> m.predict_topk(sf['image'])
+        """
+        try:
+            import graphlab as _gl
+        except ImportError:
+            import sframe as _gl
+        except ImportError:
+            raise ImportError('Require GraphLab Create or SFrame')
+
+        # Check input
+        if k > len(self.labels):
+            k = len(self.labels)
+
+        # Make DataIter
+        dataiter = self._make_dataiter(data, batch_size)
+
+        # Make prediction
+        # pred[i][j] is the score of row i belongs to label j
+        pred = self.model.predict(dataiter)
+        # top_idx[i][k] is the label index of kth highest score of row i
+        top_idx = pred.argsort()[:,-k:][:,::-1]
+        # Take row wise index, to get topk score for each row
+        top_scores = pred[_np.arange(pred.shape[0])[:,None], top_idx]
+
+        top_labels = [self.labels[i] for i in top_idx.flatten()]
+        row_ids = _np.repeat(_np.arange(len(pred)), k)
+        ranks = _np.tile(_np.arange(1, k+1), len(pred))
+
+        ret = _gl.SFrame()
+        ret['row_id'] = row_ids
+        ret['label_id'] = top_idx.flatten()
+        ret['label'] = top_labels
+        ret['score'] = top_scores.flatten()
+        ret['rank'] = ranks
+        return ret
+
+class ImageDetector(object):
+    """
+    Wrapeer of pretrained image detect model.
+
+    Use :py:func:`load_model` or :py:func:`load_path` to load the model. Do not
+    construct directly.
+
+    Parameters
+    ----------
+    model: FeedForward
+        The underlying model
+    labels : list
+        Map from index to label
+    metadata: dict
+        Metadata of the model including name, version, input shape, etc
+    """
+    def __init__(self, model, labels, metadata):
+        self._model = model
+        self._labels = labels
+        self._metadata = metadata
+        self._num_classes = len(labels)
+        # rcnn params
+        for k, v in metadata.items():
+            self.__setattr__("_" + k, v)
+
+        # new executor
+        self._ctx = model.ctx[0] # only support single device for now
+        self._sym = model.symbol
+        self._arg_params = {k : v.as_in_context(self._ctx) for k, v in model.arg_params.items()}
+        self._aux_params = {k : v.as_in_context(self._ctx) for k, v in model.aux_params.items()}
+        self._executor_with_feature = False
+
+    def _init_executor(self, sym):
+        #self._arg_params["data"] = _ndarray.zeros(self._input_shape, self._ctx)
+        self._executor = sym.bind(self._ctx, self._arg_params, args_grad=None,
+                                  grad_req="null", aux_states=self._aux_params)
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def labels(self):
+        return self._labels
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def version(self):
+        return self._version
+
+    def _detect(self, im):
```

Comment:
if the type of im is gl.Image, maybe we should use gl_im as the variable name. "im" by default refers to PIL image or NDarray image.